crates/red_knot_python_semantic/resources/mdtest/narrow/issubclass.md

```diff
@@ -0,0 +1,212 @@
+# Narrowing for `issubclass` checks
+
+Narrowing for `issubclass(class, classinfo)` expressions.
+
+## `classinfo` is a single type
+
+### Basic example
+
+```py
+def flag() -> bool: ...
+
+t = int if flag() else str
+
+if issubclass(t, bytes):
+    reveal_type(t)  # revealed: Never
+
+if issubclass(t, object):
+    reveal_type(t)  # revealed: Literal[int, str]
+
+if issubclass(t, int):
+    reveal_type(t)  # revealed: Literal[int]
+else:
+    reveal_type(t)  # revealed: Literal[str]
+
+if issubclass(t, str):
+    reveal_type(t)  # revealed: Literal[str]
+    if issubclass(t, int):
+        reveal_type(t)  # revealed: Never
+```
+
+### Proper narrowing in `elif` and `else` branches
+
+```py
+def flag() -> bool: ...
+
+t = int if flag() else str if flag() else bytes
+
+if issubclass(t, int):
+    reveal_type(t)  # revealed: Literal[int]
+else:
+    reveal_type(t)  # revealed: Literal[str, bytes]
+
+if issubclass(t, int):
+    reveal_type(t)  # revealed: Literal[int]
+elif issubclass(t, str):
+    reveal_type(t)  # revealed: Literal[str]
+else:
+    reveal_type(t)  # revealed: Literal[bytes]
+```
+
+### Multiple derived classes
+
+```py
+class Base: ...
+class Derived1(Base): ...
+class Derived2(Base): ...
+class Unrelated: ...
+
+def flag() -> bool: ...
+
+t1 = Derived1 if flag() else Derived2
+
+if issubclass(t1, Base):
+    reveal_type(t1)  # revealed: Literal[Derived1, Derived2]
+
+if issubclass(t1, Derived1):
+    reveal_type(t1)  # revealed: Literal[Derived1]
+else:
+    reveal_type(t1)  # revealed: Literal[Derived2]
+
+t2 = Derived1 if flag() else Base
+
+if issubclass(t2, Base):
+    reveal_type(t2)  # revealed: Literal[Derived1, Base]
+
+t3 = Derived1 if flag() else Unrelated
+
+if issubclass(t3, Base):
+    reveal_type(t3)  # revealed: Literal[Derived1]
+else:
+    reveal_type(t3)  # revealed: Literal[Unrelated]
+```
+
+### Narrowing for non-literals
+
+```py
+class A: ...
+class B: ...
+
+def flag() -> bool: ...
+def get_class() -> object: ...
+
+t = get_class()
+
+if issubclass(t, A):
+    reveal_type(t)  # revealed: type[A]
```

Comment:
Interestingly, this is an area where mypy and pyright differ. Pyright does the narrowing that you do here, but mypy actually refuses to do any type narrowing here as a point of principle. This is because (unlike for `isinstance()`) arbitrary objects are not acceptable as the first argument to `issubclass()`, only instances of `builtins.type`. Since `t` has type `object` here (which is not a subtype of `builtins.type`), mypy doesn't know whether or not the call will even succeed, so it doesn't:

```pycon
>>> issubclass(3, str)
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    issubclass(3, str)
    ~~~~~~~~~~^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

To be clear, both type checkers issue a diagnostic on the `if issubclass(t, A)` call warning that it's not okay to pass an arbitrary object as the first argument to `issubclass()`. But pyright then proceeds to narrow the type anyway, even though it knows the call could fail, whereas mypy refuses to.

I think I like mypy's behaviour better here? But not sure.

- https://mypy-play.net/?mypy=latest&python=3.12&gist=1db65d922577e1c2fe52bd043070d9df
- https://pyright-play.net/?strict=true&code=MYGwhgzhAECCBc0B0KBQpIwEKJU1qAJgKYBm0p4A5gBQCU0AtAHzQBGA9hyLmieVWIAXAPoYo9Jqw5sAVsWBDe%2BVEOgBeaINHiI9AgEtyBqAFc2umkIA0cOvFTQn0AE7EAbsTAgRQgJ4ADsRWDNAAxK4eXiDEhIj%2BQQDasAC6js5G0CYQ5pY20Fj26c7Obp7evoHBQqERZdGx8VXJKdAAZNAJxIlYacQgEMQOJfUVXSFOdVHejdAy8ort0AB%2BXS1AA

Mypy's behaviour means that it _additionally_ complains about the second `issubclass()` call, whereas pyright does not.

Reply:
I prefer pyright's behavior here. I think in cases where an expression may fail, but _if_ it does not fail, we can be confident of a type, it is better to emit the diagnostic and infer the "if it didn't fail" type, rather than inferring `Unknown`. This is the parallel case, but for narrowing. If the `isinstance` call fails, the code inside the `if` won't run anyway -- so why not infer the type we know must be the case if that code runs?

But perhaps more important here -- if we are going to write a test that should emit a diagnostic for invalid argument to `issubclass`, I want a TODO comment for that diagnostic. (Though here I would also be fine if we used `type[object]` instead of `object` to avoid that diagnostic.)

Reply:
> I think in cases where an expression may fail, but _if_ it does not fail, we can be confident of a type, it is better to emit the diagnostic and infer the "if it didn't fail" type, rather than inferring `Unknown`.

Okay, that makes sense for `builtins.object`, since obviously an instance of `object` _could_ be an instance of `type`, and therefore a valid first argument to `issubclass()` -- just don't _know_ that it is.

What about types that are entirely disjoint with `Instance(builtins.type)`, however -- where we _know_ that the runtime value cannot simultaneously inhabit that type and also be a valid first argument to `issubclass()`? Should we still narrow there? E.g. pyright has some "interesting" inference here...

```py
class A: ...
class B: ...

def flag() -> bool: ...
t = 3

if issubclass(t, A):
    reveal_type(t)  # Type of "t" is "type[<subclass of Literal[3] and A>]"
    if issubclass(t, B):
        reveal_type(t)  # Type of "t" is "type[<subclass of <subclass of Literal[3] and A> and B>]"
else:
    reveal_type(t)  # Type of "t" is "type[Literal[3]]"
```

"Narrowing" to types that we know cannot _ever_ exist makes little sense to me.

Reply:
Maybe we already do better than pyright on that point, though? Since intersections are a first-class concept for us when they're not for pyright?

Some tests for types that are entirely disjoint from `Instance(builtins.type)` might be great to clarify this!

Reply:
Yes, I think in those cases we would just infer `Never`, which is perfectly sensible, because our intersection handling is better.

There's a case to be made that `Never` (rather than `Unknown`) is in general a better "forgiving" type to infer from failure cases.

Reply:
> But perhaps more important here -- if we are going to write a test that should emit a diagnostic for invalid argument to `issubclass`, I want a TODO comment for that diagnostic.

Done.

> Though here I would like it even better if we used `type[object]` instead of `object` to avoid that diagnostic

Ha! I had `type` here initially, but was annoyed by how it looked in the revealed types (`type & ~type[A]`), so I changed it to object without thinking about the fact that it is too wide of a type for the first argument of `issubclasss`. Changed to `type[object]` now.



> and wrote a separate test with TODO specifically for the invalid-argument case

Done, added two new tests.

> But using `type[object]` would require enhancing `infer_type_expression` to understand the `type[T]` syntactic form.

Implemented an initial version of this.

> What about types that are entirely disjoint with `Instance(builtins.type)`, however

Added a test for this specific case as well.

> I think in those cases we would just infer `Never`, which is perfectly sensible, because our intersection handling is better.

Correct:
```py
t = 1

# TODO: we should emit a diagnostic here
if issubclass(t, int):
    reveal_type(t)  # revealed: Never
```